[Sema] Requestify default argument type checking

hamishknight opened the pull request:
This PR introduces two new requests, one to compute initializer contexts for default arguments, and another to type-check default arguments within this context.

This PR then splits `getDefaultValue` into 2 accessors:

- `getStructuralDefaultExpr` which retrieves the potentially un-type-checked default argument expression.

- `getTypeCheckedDefaultExpr` which retrieves a fully type-checked default argument expression.

And also adds `hasDefaultExpr`, which allows checking for the presence of a default expr without kicking off a request.

On its own, this isn't enough to resolve [SR-11085](https://bugs.swift.org/browse/SR-11085), I'm going to tackle that in a follow-up to keep the size of this PR down.

hamishknight committed:
[AST] Add swift::getParameterList

hamishknight committed:
Requestify default initializer context creation

This commit adds a request that computes
the initializer context for a parameter with a
default expr or stored property default.

This avoids having to compute them for synthesized
decls and is a step towards requestifying default
argument parsing.

hamishknight commented:
@swift-ci please test

CodaFi commented:
A high level comment: Why do we need to keep around a trichotomy of "parsed/typechecked/unchecked" if all of the callers before don't really care about the distinction between parsed and typechecked?  I think you can get away with "definitely typechecked" and "potentially typechecked - no requests".

hamishknight commented:
@CodaFi My understanding was that we were aiming to eventually get rid of that mutable state within the AST, so the result of `getParsedDefaultExpr` doesn't change if we happen to trigger the type-checking of the default.

CodaFi commented:
I guess I'm wondering what's stopping us from doing it now?

hamishknight commented:
@CodaFi From removing the mutable state? Things like the AST walker currently still need to set it (until it is changed to return a new AST), and as I understand it, it's currently more efficient to access the state from the AST than the evaluator's cache.

hamishknight commented:
Ping @slavapestov

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

slavapestov commented:
Sorry, I’ll review this today.

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -5208,7 +5211,8 @@ class ParamDecl : public VarDecl {
   struct StoredDefaultArgument {
-    PointerUnion<Expr *, VarDecl *> DefaultArg;
+    PointerUnion<Expr *, VarDecl *> ParsedExprOrVar;
+    Optional<Expr *> TypeCheckedExpr;
```
I'm not sure it makes sense to split these up currently. Type checking an expression mutates it, destroying the original value.

hamishknight replied:
@slavapestov Ah right, I guess in that case we should just expose something like `getStructuralDefaultExpr` and `getTypeCheckedDefaultExpr` where the former returns either an un-type-checked or a type-checked expr?

slavapestov commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -1744,6 +1744,47 @@ class ValueWitnessRequest
 
+/// Computes initializer contexts for each of the parameters with default
+/// arguments within a parameter list.
+class DefaultArgumentInitContextsRequest
```
Is it really worth making this a request? It looks like a request that's run for its side effects, which is something we would like to avoid having. Maybe it should at least be per-ParamDecl and not per-ParameterList?

hamishknight replied:
@slavapestov The reasoning for putting at the level of `ParameterList` is to allow us to access the parameter index when creating the contexts. We could make it per-ParamDecl, but then we'd need to iterate over the parameter list for each param to find its own index. Either that or start storing the index on ParamDecl. What do you think?

hamishknight replied:
@slavapestov On second thought, we can totally make this per-ParamDecl, we can just cache the other parameter's init contexts while we iterate. Does the updated code look good?

hamishknight committed:
[Sema] Requestify default arg type checking

This commit introduces a request to type-check a
default argument expression and splits
`getDefaultValue` into 2 accessors:

- `getStructuralDefaultExpr` which retrieves the
potentially un-type-checked default argument
expression.

- `getTypeCheckedDefaultExpr` which retrieves a
fully type-checked default argument expression.

In addition, this commit adds `hasDefaultExpr`,
which allows checking for a default expr without
kicking off a request.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

CodaFi commented:
Sorry in advance @hamishknight, #28179 is gonna cause a bunch of merge conflicts here.

hamishknight commented:
@CodaFi No worries, it's worth it :)

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test Linux platform

hamishknight commented:
@swift-ci please smoke test Linux platform

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -6046,15 +6046,79 @@ AnyFunctionType::Param ParamDecl::toFunctionParam(Type type) const {
 }
 
-void ParamDecl::setDefaultValue(Expr *E) {
+Initializer *ParamDecl::getDefaultArgumentInitContextCached() const {
```
This could return an Optional<> I think. Also typically we name these methods "getCached...()"

hamishknight committed:
Address review feedback

Rename getDefaultArgumentInitContextCached,
and have it return an Optional<Initializer *>.

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test
